workers_kv_test.go

```diff
@@ -0,0 +1,369 @@
+		})
+		sort.Slice(want.Result, func(i, j int) bool {
+			return want.Result[i].ID < want.Result[j].ID
+		})
+		assert.Equal(t, res.Result, want.Result)
+	}
+}
+
+func TestWorkersKV_UpdateStorageNamespace(t *testing.T) {
+	setup(UsingOrganization("foo"))
+	defer teardown()
+
+	namespace := "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+	response := `{
+		"result": null,
+		"success": true,
+		"errors": [],
+		"messages": []
+	}`
+
+	mux.HandleFunc(fmt.Sprintf("/accounts/foo/storage/kv/namespaces/%s", namespace), func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "PUT", r.Method, "Expected method 'PUT', got %s", r.Method)
+		w.Header().Set("content-type", "application/javascript")
+		fmt.Fprintf(w, response)
+	})
+
+	res, err := client.UpdateStorageNamespace(context.Background(), namespace, &StorageNamespaceRequest{Title: "Namespace"})
+	want := successResponse
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, want, res)
+	}
+}
+
+func TestWorkersKV_CreateStorageKV(t *testing.T) {
+	setup(UsingOrganization("foo"))
+	defer teardown()
+
+	key := "test_key"
+	value := []byte("test_value")
+	namespace := "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+	response := `{
+		"result": null,
+		"success": true,
+		"errors": [],
+		"messages": []
+	}`
+
+	mux.HandleFunc(fmt.Sprintf("/accounts/foo/storage/kv/namespaces/%s/values/%s", namespace, key), func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "PUT", r.Method, "Expected method 'PUT', got %s", r.Method)
+		w.Header().Set("content-type", "binary/octet-stream")
+		fmt.Fprintf(w, response)
+	})
+
+	res, err := client.CreateStorageKV(context.Background(), namespace, key, bytes.NewReader(value))
+	want := successResponse
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, want, res)
+	}
+}
+
+func TestWorkersKV_CreateStorageKVRetryResetsBody(t *testing.T) {
+	setup(UsingOrganization("foo"), UsingRetryPolicy(1, 0, 0))
+	defer teardown()
+
+	key := "test_key"
+	value := []byte("test_value")
+	namespace := "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+	response := `{
+		"result": null,
+		"success": true,
+		"errors": [],
+		"messages": []
+	}`
+
+	try := 0
+	mux.HandleFunc(fmt.Sprintf("/accounts/foo/storage/kv/namespaces/%s/values/%s", namespace, key), func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "PUT", r.Method, "Expected method 'PUT', got %s", r.Method)
+
+		res, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			t.Fatal(err)
+		}
+		assert.Equal(t, value, res)
+
+		// Fail the first time after reading the request payload
+		if try < 1 {
+			w.WriteHeader(http.StatusInternalServerError)
+			try++
+		}
+
+		w.Header().Set("content-type", "binary/octet-stream")
+		fmt.Fprintf(w, response)
+	})
+
+	reader := bytes.NewReader(value)
+	totalLen := reader.Len()
+	assert.True(t, totalLen > 0)
+
+	res, err := client.CreateStorageKV(context.Background(), namespace, key, reader)
+
+	if assert.NoError(t, err) {
+		assert.Equal(t, successResponse, res)
+	}
```

Comment:
i don't understand this, shouldn't you get an error?

Reply:
It only errors the first time. The retry logic in `makeRequestWithAuthTypeAndHeaders` calls the mock handler a second time according to the retry policy which succeeds.